RFC: pmem: WIP pmem_deep_flush

GBuella opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/2371)
<!-- Reviewable:end -->

krzycz commented:
Reviewed 19 of 22 files at r1.
Review status: 19 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[src/libpmemblk/Makefile, line 48 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOl4MzjJVDO8k_KQG0:-KzOl4MzjJVDO8k_KQG1:b-hazj39) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmemblk/Makefile#L48)):*
> ```Makefile
> include ../Makefile.inc
> 
> LIBS += -pthread -lpmem -lndctl
> ```

Why not:
`LIBS += $(shell $(PKG_CONFIG) --libs libndctl)`

---

*[src/libpmemobj/Makefile, line 70 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOl8vVTPRd6xuCqhvs:-KzOl8vVTPRd6xuCqhvt:b32v1h) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmemobj/Makefile#L70)):*
> ```Makefile
> CFLAGS += -DUSE_LIBDL -D_PMEMOBJ_INTRNL
> 
> LIBS += -pthread -lpmem $(LIBDL) -lndctl
> ```

''

---

*[src/libvmem/Makefile, line 58 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlH2NFq3iRnjjwnc6:-KzOlH2NFq3iRnjjwnc7:boiry36) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libvmem/Makefile#L58)):*
> ```Makefile
> EXTRA_OBJS += $(JEMALLOC_LIB)
> LIBS += -pthread
> LIBS += $(shell $(PKG_CONFIG) --libs libndctl)
> ```

Is this really needed? libvmem does not flush anything.

---

*[src/libvmmalloc/Makefile, line 58 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlOGzerEeuLwrfdh2:-KzOlOGzerEeuLwrfdh3:b32v1h) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libvmmalloc/Makefile#L58)):*
> ```Makefile
> EXTRA_OBJS += $(JEMALLOC_LIB)
> LIBS += -pthread
> LIBS += $(shell $(PKG_CONFIG) --libs libndctl)
> ```

''

---

*[src/test/Makefile.inc, line 190 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlRnfBqUT491RJ-0P:-KzOlRnfBqUT491RJ-0Q:b32v1h) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/test/Makefile.inc#L190)):*
> ```Makefile
> STATIC_DEBUG_LIBS += $(LIBS_DIR)/debug/libvmem.a
> STATIC_NONDEBUG_LIBS += $(LIBS_DIR)/nondebug/libvmem.a
> LIBS += $(shell $(PKG_CONFIG) --libs libndctl)
> ```

''

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

pbalcer commented on src/common/mmap.c:
```diff
@@ -521,31 +537,174 @@ util_range_is_pmem(const void *addr, size_t len)
+		return -1;
+	}
+
+	close(deep_flush_fd);
```
if you are worried about close() overwriting the errno (as evident a few lines above), do it everywhere.

pbalcer commented on src/common/mmap.c:
```diff
@@ -521,31 +537,174 @@ util_range_is_pmem(const void *addr, size_t len)
 }
+
+static int
+acquire_region_fd(int slash_dev, int regs_dir_fd, struct ndctl_ctx *ctx,
```
this is linux specific, make a generic API and implement it in a different file.

gaweinbergi commented:
*[src/libpmemblk/Makefile, line 48 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOl4MzjJVDO8k_KQG0:-KzPCxsTmV9vEooZtIXz:b-om6z8w) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmemblk/Makefile#L48)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why not:
`LIBS += $(shell $(PKG_CONFIG) --libs libndctl)`

</blockquote></details>

Also, libndctl isn't on FreeBSD yet, so this will fail on FreeBSD. Add a check_package and set a variable in common.inc? Then when libndctl becomes available on FreeBSD, no additional Makefile changes will be needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

krzycz commented:
1) missing man page
2) missing Windows support

---

Reviewed 3 of 22 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[src/common/mmap.c, line 48 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOm03WX7swJv1XEyWT:-KzOm03WX7swJv1XEyWU:b5kv3o3) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/common/mmap.c#L48)):*
> ```C
> #include <sys/mman.h>
> #include <unistd.h>
> #include <ndctl/libndctl.h>
> ```

That would break FreeBSD, I'm afraid.

---

*[src/libpmem/pmem.c, line 665 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOn0mgGCvTvm2fOj9j:-KzOn0mhTu6_lYNML9DI:b7ux8xf) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmem/pmem.c#L665)):*
> ```C
> #ifndef _WIN32
> 	/* XXX only Device DAX regions (PMEM) are tracked so far */
> 	if (is_dev_dax && util_range_register(addr, len, fd) != 0) {
> ```

Storing *fd* on the mapping list is pointless, as *fd* is closed a few lines below.
You'd need to duplicate the file descriptor and close it when the associated memory region is unmapped. 
BTW, this is how it is implemented on Windows.

---

*[src/libpmem/pmem.c, line 1283 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOnZSKeuJTAL_WuFcW:-KzOnZSKeuJTAL_WuFcX:b-6oleat) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmem/pmem.c#L1283)):*
> ```C
> 
> int
> pmem_deep_flush(const void *addr, size_t len)
> ```

missing brief

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371#-:-KzOndhMIiHiIfrsd2Sa:b-bgw1tl)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[src/libvmem/Makefile, line 58 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlH2NFq3iRnjjwnc6:-KzTUovXVC7-QgzzZUvv:bo5758) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libvmem/Makefile#L58)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is this really needed? libvmem does not flush anything.
</blockquote></details>

The code I used to track memory regions is in src/common/mmap.c, which is linked into libvmem. Without using libndctl:
```
cc -o arttree arttree.o -Wl,-rpath=../../../debug -L../../../debug  -Wl,-rpath=. -L. -lart -lvmem
../../../debug/libvmem.so: undefined reference to `ndctl_bus_get_next'
../../../debug/libvmem.so: undefined reference to `ndctl_new'
../../../debug/libvmem.so: undefined reference to `ndctl_region_get_next'
../../../debug/libvmem.so: undefined reference to `ndctl_region_get_devname'
../../../debug/libvmem.so: undefined reference to `ndctl_region_get_first'
../../../debug/libvmem.so: undefined reference to `ndctl_unref'
../../../debug/libvmem.so: undefined reference to `ndctl_bus_get_first'
../../../debug/libvmem.so: undefined reference to `ndctl_dax_get_devname'
../../../debug/libvmem.so: undefined reference to `ndctl_dax_get_next'
../../../debug/libvmem.so: undefined reference to `ndctl_dax_get_first'
collect2: error: ld returned 1 exit status
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[src/libvmem/Makefile, line 58 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlH2NFq3iRnjjwnc6:-KzTYE50gFQzo_zLnr9l:b-l27d5a) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libvmem/Makefile#L58)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

The code I used to track memory regions is in src/common/mmap.c, which is linked into libvmem. Without using libndctl:
```
cc -o arttree arttree.o -Wl,-rpath=../../../debug -L../../../debug  -Wl,-rpath=. -L. -lart -lvmem
../../../debug/libvmem.so: undefined reference to `ndctl_bus_get_next'
../../../debug/libvmem.so: undefined reference to `ndctl_new'
../../../debug/libvmem.so: undefined reference to `ndctl_region_get_next'
../../../debug/libvmem.so: undefined reference to `ndctl_region_get_devname'
../../../debug/libvmem.so: undefined reference to `ndctl_region_get_first'
../../../debug/libvmem.so: undefined reference to `ndctl_unref'
../../../debug/libvmem.so: undefined reference to `ndctl_bus_get_first'
../../../debug/libvmem.so: undefined reference to `ndctl_dax_get_devname'
../../../debug/libvmem.so: undefined reference to `ndctl_dax_get_next'
../../../debug/libvmem.so: undefined reference to `ndctl_dax_get_first'
collect2: error: ld returned 1 exit status
```
</blockquote></details>

I see. So, it's a good opportunity to refactor this code. Or, even if this file is shared between libpmem\* and libvmem\* libraries, we can have some \#ifdef's to disable ndctl-dependant  code when compiled for libvmem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 9 of 14 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/common/mmap.c, line 48 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOm03WX7swJv1XEyWT:-KzTxs6qagRDGh4viILx:b-iwiyft) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/common/mmap.c#L48)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

That would break FreeBSD, I'm afraid.
</blockquote></details>

Moved to a separate file.

---

*[src/common/mmap.c, line 573 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzNwKIPZT4hEeW-ugYA-r1-573:-KzTxx80gDZeN6sjsBeM:bptcuqh) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/common/mmap.c#L573)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

this is linux specific, make a generic API and implement it in a different file.
</blockquote></details>

Moved to a different file.

---

*[src/libpmem/pmem.c, line 665 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOn0mgGCvTvm2fOj9j:-KzTy1PJjmbxunykXujG:bg7ai0t) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmem/pmem.c#L665)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Storing *fd* on the mapping list is pointless, as *fd* is closed a few lines below.
You'd need to duplicate the file descriptor and close it when the associated memory region is unmapped. 
BTW, this is how it is implemented on Windows.
</blockquote></details>

I store the dev_id instead.

---

*[src/libvmem/Makefile, line 58 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlH2NFq3iRnjjwnc6:-KzTy9IbpbU-LmAyGGex:bddh3o) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libvmem/Makefile#L58)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

I see. So, it's a good opportunity to refactor this code. Or, even if this file is shared between libpmem\* and libvmem\* libraries, we can have some \#ifdef's to disable ndctl-dependant  code when compiled for libvmem.
</blockquote></details>

I use #ifdef for this now, as suggested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

gaweinbergi commented:
*[src/test/Makefile.inc, line 190 at r2](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzVRIACl-CszcfJXhor:-KzVRIACl-CszcfJXhos:b-hj89pp) ([raw file](https://github.com/pmem/nvml/blob/a1b090547046af144b2cfeedb0136b7584271fd1/src/test/Makefile.inc#L190)):*
> ```Makefile
> STATIC_DEBUG_LIBS += $(LIBS_DIR)/debug/libvmem.a
> STATIC_NONDEBUG_LIBS += $(LIBS_DIR)/nondebug/libvmem.a
> LIBS += $(shell $(PKG_CONFIG) --libs libndctl)
> ```

This generates a ton of notices when libndctl is not present (e.g., on FreeBSD). Suggest adding a check_package in common.inc and setting a variable (similar to LIBDL) that is used in the subsidiary Makefiles. Your choice whether you generate a single warning or just ignore the lack of libndctl (since the code is designed to build without it).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

gaweinbergi commented:
*[src/libpmem/Makefile, line 53 at r2](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzVSEIo-7qAQnc7zUP-:-KzVOQIrp2cIWQx5-9og:bw9wuon) ([raw file](https://github.com/pmem/nvml/blob/a1b090547046af144b2cfeedb0136b7584271fd1/src/libpmem/Makefile#L53)):*
> ```Makefile
> 	pmem_linux.c
> 
> ifeq ($(shell uname -s),Linux)
> ```

Why not check for libndctl using check_package instead? That way once libndctl is supported on FreeBSD, everything will just work without requiring any changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 7 of 15 files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

*[src/libpmem/Makefile, line 53 at r2](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzVSEIo-7qAQnc7zUP-:-KzYN8E2OU5RtEVtAcpl:blqvkqs) ([raw file](https://github.com/pmem/nvml/blob/a1b090547046af144b2cfeedb0136b7584271fd1/src/libpmem/Makefile#L53)):*
<details><summary><i>Previously, gaweinbergi (Glenn Weinberg) wrote…</i></summary><blockquote>

Why not check for libndctl using check_package instead? That way once libndctl is supported on FreeBSD, everything will just work without requiring any changes.
</blockquote></details>

I use check_package now.
The main question that remains, is when/on what platforms do we allow building without libndctl?
I guess on Linux it should be required, since deep_flush would be (silently) erroneous on DAX devices without it.
What about FreeBSD? Will it ever implement a similar feature? Should we allow building without libndctl now, but not with some later versions of FreeBSD?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 8 of 15 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

*[src/common/ddax_range_deep_flush.c, line 34 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYrwwZxplIGwcRqiI5:-KzYrwwZxplIGwcRqiI6:b-npkx4g) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L34)):*
> ```C
> 
> /*
>  * ddax_range_deep_flush.c -- mmap utilities
> ```

Pls, indicate this file is Linux-specific.

---

*[src/common/ddax_range_deep_flush.c, line 47 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYrVJr-G_zCv087I4i:-KzYrVJr-G_zCv087I4j:b-u7nyoz) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L47)):*
> ```C
> #include "ddax_range_deep_flush.h"
> 
> static int
> ```

missing briefs
missing debug trace points (LOG/ERR macros)

---

*[src/common/ddax_range_deep_flush.c, line 52 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYsitRuaFj_lVlE85k:-KzYsitRuaFj_lVlE85l:b-9lga9) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L52)):*
> ```C
> {
> 	struct ndctl_bus *bus;
> 	ndctl_bus_foreach(ctx, bus) {
> ```

So, each time the program would like to issue a deep flush, we will iterate thru all the buses/regions/devices to find the right one?
Why not to determine (open?) the associated "deep_flush" file location once, when the device is mapped to memory / the memory range is registered? Do you assume the sysfs entry point may change in the meantime?

---

*[src/common/ddax_range_deep_flush.c, line 86 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYs9td7ikoOPddDFA1:-KzYs9td7ikoOPddDFA2:buhuwce) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L86)):*
> ```C
> 	int retval = -1;
> 	struct ndctl_ctx *ctx = NULL;
> 	int regs_dir = -1;
> ```

Any reason for initializing these variables?

---

*[src/common/ddax_range_deep_flush.c, line 97 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzZ8XNGEU2hGWwKqwxb:-KzZ8XNGEU2hGWwKqwxc:bmw2r7h) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L97)):*
> ```C
> 		goto done;
> 	region_fd =
> 	    acquire_region_fd(slash_dev, regs_dir, ctx, dev_id);
> ```

A very naive question:
Since the device dax name (id) is constructed from its parent region id (see: "linux/drivers/dax/device.c") , then for "/dev/dax7.0", the sysfs entry for its parent region is "/sys/bus/nd/devices/region7", isn't it?
Knowing that, it seems for me that finding the associated "deep_flush" file is trivial - just take the real device file name, extract the region id, etc...
IOW, having *dev_id*, read the file "/sys/dev/char/X.Y/device/dax_region/id" (where X/Y = major/minor(dev_id)) to get the region id (R). Then write '1" to the file "/sys/bus/nd/devices/regionR". That's it.
Do we really need to use libndctl for it? Am I missing something?

---

*[src/common/mmap.c, line 83 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYdj4A19dEq0cHGj6H:-KzYdj4A19dEq0cHGj6I:b-ternso) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L83)):*
> ```C
> #endif
> 
> 	dev_t dev_id;
> ```

\#ifndef \_WIN32 ? (there's no Device DAX on Windows)

---

*[src/common/mmap.c, line 393 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYdtiKUijMTZdHs5HP:-KzYdtiKUijMTZdHs5HQ:bonrt3) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L393)):*
> ```C
> 
> 	struct stat fd_stat;
> 	if (fstat(fd, &fd_stat) != 0) {
> ```

1) os_fstat / os_stat_t
2) \#ifndef \_WIN32

---

*[src/common/mmap.c, line 406 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYeHmudweSX9azXGhY:-KzYeHmudweSX9azXGhZ:bsm77ix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L406)):*
> ```C
> 	SORTEDQ_INSERT(&Mmap_list, mt, entry, struct map_tracker,
> 			util_range_comparer);
> ```

`return 0;`
Otherwise, it would always free the list entry before return.

---

*[src/common/mmap.c, line 588 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYfHJ8Ul5dCr1V81xN:-KzYfHJ8Ul5dCr1V81xO:b-940r89) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L588)):*
> ```C
> {
> #ifdef NVML_COMMON_MMAP_DAX_DEEP_FLUSH
> 	const struct map_tracker *mt = util_range_find(addr, len);
> ```

I'd prefer `while (len > 0) {` over recursion, but this is also fine.

---

*[src/common/mmap.c, line 621 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYqWnzO_9cPanQdKv-:-KzYqWnzO_9cPanQdKv0:b-6iu3to) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L621)):*
> ```C
> util_range_deep_flush(const void *addr, size_t len)
> {
> 	LOG(10, "addr %p len %zu", addr, len);
> ```

Why not 3? It won't be called very often, I suppose.

---

*[src/libpmem/Makefile, line 53 at r2](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzVSEIo-7qAQnc7zUP-:-KzYhvRzCuN-FtpiDGpR:bz0jz0j) ([raw file](https://github.com/pmem/nvml/blob/a1b090547046af144b2cfeedb0136b7584271fd1/src/libpmem/Makefile#L53)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

I use check_package now.
The main question that remains, is when/on what platforms do we allow building without libndctl?
I guess on Linux it should be required, since deep_flush would be (silently) erroneous on DAX devices without it.
What about FreeBSD? Will it ever implement a similar feature? Should we allow building without libndctl now, but not with some later versions of FreeBSD?
</blockquote></details>

I hope Device DAX will **never** be supported on FreeBSD. :-)
If there will be any device providing raw access to persistent memory (w/o the intervening filesystem), it should be implemented in a way that allows to issue a regular syscall (i.e. msync) to perform a deep flush.

---

*[src/libpmem/pmem.c, line 1285 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYq1BPEcLFOWBb9U1X:-KzYq1BPEcLFOWBb9U1Y:b-oevgjh) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/libpmem/pmem.c#L1285)):*
> ```C
> pmem_deep_flush(const void *addr, size_t len)
> {
> 	return util_range_deep_flush(addr, len);
> ```

missing `LOG(3, ...)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

gaweinbergi commented:
*[src/libpmem/Makefile, line 53 at r2](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzVSEIo-7qAQnc7zUP-:-KzZha5USs2imZ0tpmss:b-wq4td2) ([raw file](https://github.com/pmem/nvml/blob/a1b090547046af144b2cfeedb0136b7584271fd1/src/libpmem/Makefile#L53)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

I hope Device DAX will **never** be supported on FreeBSD. :-)
If there will be any device providing raw access to persistent memory (w/o the intervening filesystem), it should be implemented in a way that allows to issue a regular syscall (i.e. msync) to perform a deep flush.
</blockquote></details>

If you have specs for how you would like it to work, let me know and I’ll pass them on to the FreeBSD kernel developer. He’s working on NVDIMM-N support right now. And/or I can make introductions. He’s in the Ukraine :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 8 of 22 files at r1, 7 of 15 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks broke.

---

*[src/common/ddax_range_deep_flush.c, line 34 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYrwwZxplIGwcRqiI5:-KzcNc4ll_u1gfqdZa8t:b-auxg3t) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L34)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Pls, indicate this file is Linux-specific.
</blockquote></details>

And there is no mmap utilities in this file.

---

*[src/common/ddax_range_deep_flush.c, line 52 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYsitRuaFj_lVlE85k:-KzcR-2QL70qdzhWDnwa:b-st9smo) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L52)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

So, each time the program would like to issue a deep flush, we will iterate thru all the buses/regions/devices to find the right one?
Why not to determine (open?) the associated "deep_flush" file location once, when the device is mapped to memory / the memory range is registered? Do you assume the sysfs entry point may change in the meantime?
</blockquote></details>

I would suggest to open the appropriate "deep_flush" file and keep it in the 'struct map_tracker'

---

*[src/common/ddax_range_deep_flush.c, line 86 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYs9td7ikoOPddDFA1:-KzcOxDBC1wDPu7QQKz9:btlhrw8) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L86)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Any reason for initializing these variables?
</blockquote></details>

for error handling. Which is quite different than what you can encounter in our other files. The error handling is done using just one label, that is why you need to initialize these variables in order to figure out if the file has been opened or not. I would suggest to refactor this and have multiple labels - one per each allocated resource.

In addtion I would suggest to use ERR macro in order to indicate what was the problem with performing the deep_flush - a user can use the pmem_errormsg() function.

---

*[src/common/ddax_range_deep_flush.c, line 97 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzZ8XNGEU2hGWwKqwxb:-KzcRKqNOsBAd1MMB-u6:b-4n3v26) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L97)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

A very naive question:
Since the device dax name (id) is constructed from its parent region id (see: "linux/drivers/dax/device.c") , then for "/dev/dax7.0", the sysfs entry for its parent region is "/sys/bus/nd/devices/region7", isn't it?
Knowing that, it seems for me that finding the associated "deep_flush" file is trivial - just take the real device file name, extract the region id, etc...
IOW, having *dev_id*, read the file "/sys/dev/char/X.Y/device/dax_region/id" (where X/Y = major/minor(dev_id)) to get the region id (R). Then write '1" to the file "/sys/bus/nd/devices/regionR". That's it.
Do we really need to use libndctl for it? Am I missing something?
</blockquote></details>

IMHO, we should use libndctl because 1) it's already implemented 2) it is maintained and synchronized with kernel changes.

---

*[src/common/ddax_range_deep_flush.c, line 119 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzcOd15yabNrPq9xBwe:-KzcOd16nC9FvnfXC0ZL:bsj1bzj) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L119)):*
> ```C
> 
> /*
>  * ddax_range_deep_flush -- perform deep flush of given address range,
> ```

there is no address range in arguments provided

---

*[src/common/ddax_range_deep_flush.h, line 34 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzcRbpN6TzyjSo_ESAH:-KzcRbpOnq2vTDgUtdrH:bfyxfkm) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.h#L34)):*
> ```objc
> 
> /*
>  * ddax_range_deep_flush.c -- mmap utilities
> ```

please update  the comment

---

*[src/common/mmap.c, line 370 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzcRsdNmrg14K3XYdrM:-KzcRsdNmrg14K3XYdrN:bdrb7ch) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L370)):*
> ```C
> 	int ret = 0;
> 
> 	if (fd < 0) {
> ```

I would say ASSERT - this is internal API and there is no point in handlin this error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks broke.

---

*[src/common/ddax_range_deep_flush.c, line 97 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzZ8XNGEU2hGWwKqwxb:-KzcbnVByC-kjlPlzETJ:b-5p1zyf) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L97)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

IMHO, we should use libndctl because 1) it's already implemented 2) it is maintained and synchronized with kernel changes.
</blockquote></details>

I would go further and would implement `ndctl_dax_deep_flush`, submit that patch and use that function here.

@krzycz: It's probably not a guaranteed to work in all cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks broke.

---

*[src/common/ddax_range_deep_flush.c, line 97 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzZ8XNGEU2hGWwKqwxb:-Kzce38DVOCtJtWJSMEE:bmtpijg) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L97)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

I would go further and would implement `ndctl_dax_deep_flush`, submit that patch and use that function here.

@krzycz: It's probably not a guaranteed to work in all cases.
</blockquote></details>

The idea was to not rely on those paths (rely on paths as little as possible), thinking that the libndctl API is less likely to change than the Linux kernels sysfs implementation, and considering the unlikely case of deep flushing on other kernels.
If we are confident those paths are future proof enough, sure, why not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks broke.

---

*[src/common/ddax_range_deep_flush.c, line 86 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYs9td7ikoOPddDFA1:-KzchkqxklpNQ0j3ioNP:bkcmety) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L86)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

for error handling. Which is quite different than what you can encounter in our other files. The error handling is done using just one label, that is why you need to initialize these variables in order to figure out if the file has been opened or not. I would suggest to refactor this and have multiple labels - one per each allocated resource.

In addtion I would suggest to use ERR macro in order to indicate what was the problem with performing the deep_flush - a user can use the pmem_errormsg() function.
</blockquote></details>

Ah, right! So, only *slash_dev* could be left uninitialized.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 12 of 15 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[src/common/ddax_range_deep_flush.c, line 34 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYrwwZxplIGwcRqiI5:-KzcjiR-_b1qpytqO9Id:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L34)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

And there is no mmap utilities in this file.
</blockquote></details>

Done.

---

*[src/common/mmap.c, line 370 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzcRsdNmrg14K3XYdrM:-Kzcjrqj1b061NdKNBhb:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L370)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

I would say ASSERT - this is internal API and there is no point in handlin this error.
</blockquote></details>

Done.

---

*[src/common/mmap.c, line 406 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYeHmudweSX9azXGhY:-Kzcju70Ncn47emrpU5u:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L406)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

`return 0;`
Otherwise, it would always free the list entry before return.
</blockquote></details>

Done.

---

*[src/libpmem/pmem.c, line 1283 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOnZSKeuJTAL_WuFcW:-Kzck0eLpNYNt2uHAUOq:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmem/pmem.c#L1283)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing brief
</blockquote></details>

Done.

---

*[src/test/Makefile.inc, line 190 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlRnfBqUT491RJ-0P:-Kzck2TAg3Vk5kr2pFJw:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/test/Makefile.inc#L190)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---

*[src/test/Makefile.inc, line 190 at r2](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzVRIACl-CszcfJXhor:-Kzck2g0nPvkTAiE-4DJ:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a1b090547046af144b2cfeedb0136b7584271fd1/src/test/Makefile.inc#L190)):*
<details><summary><i>Previously, gaweinbergi (Glenn Weinberg) wrote…</i></summary><blockquote>

This generates a ton of notices when libndctl is not present (e.g., on FreeBSD). Suggest adding a check_package in common.inc and setting a variable (similar to LIBDL) that is used in the subsidiary Makefiles. Your choice whether you generate a single warning or just ignore the lack of libndctl (since the code is designed to build without it).
</blockquote></details>

Done.

---

*[src/libpmemblk/Makefile, line 48 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOl4MzjJVDO8k_KQG0:-Kzck3v5UgOk2rtltvtU:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmemblk/Makefile#L48)):*
<details><summary><i>Previously, gaweinbergi (Glenn Weinberg) wrote…</i></summary><blockquote>

Also, libndctl isn't on FreeBSD yet, so this will fail on FreeBSD. Add a check_package and set a variable in common.inc? Then when libndctl becomes available on FreeBSD, no additional Makefile changes will be needed.
</blockquote></details>

Done.

---

*[src/libpmemobj/Makefile, line 70 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOl8vVTPRd6xuCqhvs:-Kzck4SvNu3td8OYLoJP:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libpmemobj/Makefile#L70)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---

*[src/libvmmalloc/Makefile, line 58 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzOlOGzerEeuLwrfdh2:-Kzck58i1NVQBSluG5Js:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/libvmmalloc/Makefile#L58)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 11 of 15 files reviewed at latest revision, 23 unresolved discussions.

---

*[src/common/ddax_range_deep_flush.h, line 34 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzcRbpN6TzyjSo_ESAH:-KzcpdkSF9zaTfR1LCRG:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.h#L34)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

please update  the comment
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 11 of 15 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[src/common/ddax_range_deep_flush.c, line 47 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYrVJr-G_zCv087I4i:-Kzd3_7i5Wcpupi3xA3h:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L47)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing briefs
missing debug trace points (LOG/ERR macros)
</blockquote></details>

Done.

---

*[src/common/ddax_range_deep_flush.c, line 119 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzcOd15yabNrPq9xBwe:-Kzd3d9Mdz3oHwIwCYhr:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/ddax_range_deep_flush.c#L119)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

there is no address range in arguments provided
</blockquote></details>

Done.

---

*[src/common/mmap.c, line 83 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYdj4A19dEq0cHGj6H:-Kzd3hGzjE5EHMZoXSw8:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L83)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

\#ifndef \_WIN32 ? (there's no Device DAX on Windows)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 9 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/common/mmap.c, line 680 at r1](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzNwKIPZT4hEeW-ugYA-r1-680:-L-1Q1pLv4Bv6gUhpKDy:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/b6bc3ab328391858d258d80b7eae3bcb29c1a764/src/common/mmap.c#L680)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

if you are worried about close() overwriting the errno (as evident a few lines above), do it everywhere.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 9 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/common/mmap.c, line 393 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYdtiKUijMTZdHs5HP:-L-1Qt1nitRhpsLnyI8t:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L393)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

1) os_fstat / os_stat_t
2) \#ifndef \_WIN32
</blockquote></details>

Done.

---

*[src/common/mmap.c, line 588 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYfHJ8Ul5dCr1V81xN:-L-1Qv-YZo4yILcvRwy5:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L588)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

I'd prefer `while (len > 0) {` over recursion, but this is also fine.
</blockquote></details>

Done.

---

*[src/common/mmap.c, line 621 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYqWnzO_9cPanQdKv-:-L-1QySruPn5PueW0mBC:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/common/mmap.c#L621)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why not 3? It won't be called very often, I suppose.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 9 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/libpmem/pmem.c, line 1285 at r3](https://reviewable.io:443/reviews/pmem/nvml/2371#-KzYq1BPEcLFOWBb9U1X:-L-1RWeVLtPFn4oN7A4z:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a3416a3ff5fc2305bcdfab447cf4488e2c60e289/src/libpmem/pmem.c#L1285)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing `LOG(3, ...)`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Travis build failed:
```
src/test/obj_list/Makefile:1: error: wrong copyright date: (is: 2015-2016, should be: 2015-2017)
```

pbalcer commented on src/common/ddax_range_deep_flush.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <sys/types.h>
+
+int ddax_range_deep_flush(dev_t dev_id);
```
this should probably get integrated with the abstraction layer from #2388. I think we can do that later, once that patch is merged.

pbalcer commented on src/common/mmap.c:
```diff
@@ -521,31 +548,112 @@ util_range_is_pmem(const void *addr, size_t len)
 	return retval;
 }
+
+#ifdef NVML_COMMON_MMAP_DAX_DEEP_FLUSH
```
You should use your ddax_range_deep_flush.h to provide the abstraction layer and just implement empty/msync functions on unsupported platforms.

GBuella replied:
I was thinking about that, but the range tracking is implemented in this file, so I figured it would be a bit less intrusive change this way.

krzycz commented:
Reviewed 1 of 4 files at r4, 5 of 5 files at r5, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks broke.

---

*[src/common/mmap.c, line 588 at r7](https://reviewable.io:443/reviews/pmem/nvml/2371#-L-2kc4WbRuqShkAHFid:-L-2kc4WbRuqShkAHFie:bap1r8u) ([raw file](https://github.com/pmem/nvml/blob/568a83e8f54b4cba4bc73ee03335bde3bd1e406d/src/common/mmap.c#L588)):*
> ```C
>  */
> static int
> msync_range(uintptr_t addr, size_t len)
> ```

This is almost a copy&paste from *pmem_msync* (pmem.c).
It made me realize that we have most of the *deep_flush* implementation in *common*, not in *libpmem*. Do we plan to call *util_range_deep_flush* from anywhere else than libpmem?
Does it need to be shared among all the libraries?
I guess it's because the *map_tracker* and the corresponding functions are here, but this should be refactored and moved to libpmem entirely, as it's used only there.

---

*[src/libpmem/pmem.c, line 1285 at r7](https://reviewable.io:443/reviews/pmem/nvml/2371#-L-2eUPmij_KeIgBMp2A:-L-2eUPmij_KeIgBMp2B:b-jj452x) ([raw file](https://github.com/pmem/nvml/blob/568a83e8f54b4cba4bc73ee03335bde3bd1e406d/src/libpmem/pmem.c#L1285)):*
> ```C
>  * pmem_deep_flush -- perform deep flush on a memory range
>  *
>  * It merely acts as wrapper around an msync call in most cases, the only
> ```

To be honest, my understanding of "deep_flush" is that it includes flushing of CPU cache and iMC buffers (WPQs).
Assuming *range_deep_flush* does only flush the WPQ's, I think *pmem_deep_flush* should be implemented like this (pseudocode):
```c
if (is_dev_dax(addr, len)) {
    pmem_persist(addr, len);
    pmem_wpq_flush(get_device(addr, len)); // aka util_range_deep_flush
} else {
    pmem_msync(addr, len);
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2371)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Continued in #2468
